Create 20220720-per-certificate-owner-ref.md (#5324)
cert-manager/cert-manager: design/20220720-per-certificate-owner-ref.md

- [ ] User-facing documentation has been PR-ed against the release branch in [cert-manager/website](https://github.com/cert-manager/website)

## Summary

The flag `--enable-certificate-owner-ref` allows you to configure cert-manager to delete Secret resources when the associated Certificate is removed. 

We propose to introduce the same setting at the Certificate level so that users of the Certificate resource can decide whether the Secret resource should be removed or not.

## Use-cases

**Use-case 1: managed cert-manager installations**

[Flant](https://flant.com) manages Kubernetes clusters for their customers. The installation of cert-manager is managed by Flant. Flant uses `--enable-certificate-owner-ref=false` to lower the chance of outages of their managed components. On the other hand, customers are relying on long-lived “dev” namespaces in which they install and uninstall their applications over and over with random names. The Certificate resources are correctly removed, but the Secret resources stay and accumulate.

Source: https://github.com/deckhouse/deckhouse/pull/1601

> 

Reviewer: I object to this proposal because it does not solve the stated problem and because there are ways to solve this problem more generally, which do not require changes to cert-manager.

### Alternatives

The problem described in this use-case is that the "customers" are creating too many resources in their allocated namespace; so many that they are overloading the Flant Kubernetes API server.

One good answer to that problem is to impose an [Object Count Resource Quota](https://kubernetes.io/docs/concepts/policy/resource-quotas/#object-count-quota) which presumably would cause cert-manager to attach a meaningful "quota exceeded" error message to the `Certificate` resource. 
The developer would then know that they had to clean up stale resources from their dev namespace.

If I were such a developer I would attach a label to all the resources that I create during a given deployment,
and then delete all those resources using a label selector (`kubectl delete all --namespace=dev --label-selector=deployment-test-id=*` ).

Or if I were such a developer I would like to able to recreate a pristine dev namespace, perhaps by adding a label to my namespace. (e.g. `kubectl label ns dev recreate=daily`)

Or Flant could automatically delete any resources in "dev" namespaces which are older than X days.

Or they could implement an admission controller which forced developers to label all resources with a `time-to-live` label, and then delete any which outlived that duration.

My point is that there are many ways to solve this problem generally.

### Adding a `Certificate.Spec.CleanupPolicy` field does not **alone** solve the problem.

How will a new developer with a newly created "dev" namespace, know that it is their orphaned Secret resources that are causing the platform to become slow / unreliable?

How will that developer learn about this new field in the cert-manager Certificate resource?

And will they realise that it should be set to `cleanupPolicy: Never` in production and `cleanupPolicy: Always` in their "dev" namespaces?

If there were a way to add this field automatically to Certificate resources in certain namespaces, then I might be more convinced, but it still would not solve the general problem of "too many resources slowing down the K8S API server"

### Adding a `Certificate.Spec.CleanupPolicy` field does not solve **general** problem.

What happens when developer decides to use another operator that creates un-owned resources?
Won't they then have to find a general solution to this problem?

What happens if a developer writes a shell script that creates too many Secrets?

Reviewer 2: @wallrj hello, you are right, it is many possible ways to solve problem of orphaned certs. But cert-manager alreeady have general way to do it - enable certificate owner reference. But it is not flexible - you cannot manage this per-certificate. We solve this problem by enabling global flag --certificate-owner-ref=true and disable owner reference in certificate CR for system certificates, so developer secrets removed after removing appropriate certificates, and system certificate secrets never removed. In another cases can be handy to disable global certificate owner ref parameter and enable it on some certificates. My point is that if it is possible to set the owner of the certificate ref, it should be flexible.

Reviewer: I see your point but my (controversial) view is that the `--certificate-owner-ref` feature was a mistake,
and I think it should be deprecated and removed....**not** promoted and made more flexible.

I am strongly in favour of documenting how users can find Secrets that:
1. Have labels suggesting that they are (or were) managed by cert-manager.
2. No longer have a corresponding Certificate resource.
3. Are not mounted in any Pod.
4. Are not referenced by any Ingress.

I am in favour of a `cmctl experimental find-orphaned-secrets` command (or something similar) that performs some or all of those steps.

Reviewer 2: I understand, yes, it's a different view of the problem. We can close this PR and deprecate `--certificate-owner-ref` option.

[Flant](https://flant.com) manages large multi-tenant Kubernetes clusters. The installation of cert-manager is managed by Flant, and customers cannot edit cert-manager's configuration. Customers have access to a "prod" cluster and a "dev" cluster. On both clusters, Flant uses `--enable-certificate-owner-ref=false` to lower the chance of outages of their managed components such as the ingress controller.

On the "dev" cluster, customers are given long-lived namespaces in which they install and uninstall their applications over and over with random names, including Certificate resources. With hundreds of customers deploying approximately ten times a day to the "dev" cluster, the Secret resources that are left over by cert-manager accumulate (around 10,000 Secret resources after a few months), and the Kubernetes API becomes slow, with people having to wait for 10 seconds to list the secrets in a given namespace.

To solve this problem, Flant aims at using `cleanupPolicy: Never` on the certificates used for their managed components and use `--default-certificate-cleanup-policy=OnDelete` for the rest of the Certificates. Users won't have to change their Certificate resources.

> On the "prod" cluster, Flant recommends customers to keep the Secret resource on removal to lower the risk of outages. Flant aims to use `--default-certificate-cleanup-policy=Never` for the "prod" cluster and also aims to document the reason for this difference between "prod" and "dev".

Author: @wallrj PTAL

Reviewer: I can see that the revised story explains why it will not be necessary for the "Customers" to add the cleanupPolicy to each of their Certificate resources.

But my original comment still stands. 

1. This proposed new feature does not address the general problem of "Customers are able to create more resources than the platform API server can handle".
2. Customers could still accidentally bring down the Flant API service by accidentally (or deliberately) creating too many resources.
  1.  Why not instead add a resource count quota?
4. The mission critical Secret resources in the Flant "dev" cluster could still be accidentally deleted.
  1. Wouldn't it be better to implement a solid /  standard mechanism for preventing this situation. (see https://www.linkedin.com/pulse/kubernetes-protecting-resources-from-deletion-them-rodrigo)
5. "Customers" will now see inconsistent behaviour between their "dev" and "production" deployments, in dev their Secrets will be garbage collected and in Prod they will be orphaned. How will they learn about this?

These are only my opinions, and you might find other cert-manager maintainers who are more sympathetic to this new feature.

@maelvls What do you think? Is this how you would solve Flant's problem if you were DevOps there?

| If... | then they should replace it with... |
|-----|-----|
| `--enable-certificate-owner-ref` not passed to the controller | No change needed |
| `--enable-certificate-owner-ref=false` | Replace with `--default-secret-cleanup-policy=Never` |
| `--enable-certificate-owner-ref=true` | Replace with `--default-secret-cleanup-policy=OnDelete` |

## Design Details

cert-manager would have to change in a few places.

**Mutating webhook**

We propose to have no "value defaulting" for `cleanupPolicy` because the
"empty" value has a meaning for us: when `cleanupPolicy` is empty, the
presence or not of the flag `--enable-certificate-owner-ref` takes over.
To give more context, some other resources, such as the Pod resource,
will mutate the object when the value is "empty", for example the
`imagePullPolicy` value will default to `IfNotPresent`.

**PostIssuancePolicyChain**

In ([policies.go#L95](https://github.com/cert-manager/cert-manager/blob/b78af1ef867f8776715cae3dd6a8b83049c4d9b2/internal/controller/certificates/policies/policies.go#L95-L104)), cert-manager does a few sanity checks right after the issuer (either an

> internal or an external issue) has filled the CertificateRequest's status

Reviewer 3: ```suggestion
internal or an external issuer) has filled the CertificateRequest's status
```
?

and checks that the owner reference on the Secret resource matches the one
on the Certificate resource.

### Test Plan

- Unit tests for the changes in the secret manager controller.
- Integration tests (either fake client or envtest) checking various API behaviours.

### Graduation Criteria

We propose to release this feature in GA immediately and skip the "beta"
phase that consists of gathering user feedback, since this feature has a
low user-facing surface. We think that we will be able to take a good
decision (e.g., the name of the new field, whether it is a boolean or a
string, and which values the field can take) while developing the feature
in the PR.

We don't think this feature needs to be [feature gated][feature gate].

[feature gate]: https://cert-manager.io/docs/installation/featureflags/#list-of-current-feature-gates

### Upgrade / Downgrade Strategy

Upgrading from a version without this feature to a version with this
feature won't be breaking.

Downgrading, however, will be a breaking change, since a new field will be

> introduced.

Reviewer 3: The wording here seems wrong / inconsistent with how this section is written up in similar documents.

Author: That's a good point. Thanks!

Author: I've changed it to:

> Downgrading requires manual intervention: removing the new flag
> `--default-secret-deletion-policy` from the Deployment, adding the corresponding
> `--enable-certificate-owner-ref` and emptying the `deletionPolicy` field from
> every Certificate in the cluster.
>

### Supported Versions

This feature will be supported in all the versions of Kubernetes that are supported by cert-manager.

## Alternatives

**CSI driver**

> It is possible to use the
> [`csi-driver`](https://github.com/cert-manager/csi-driver) to circumvent
> the problem of "too many ephemeral Secret resources stored in etcd". Using
> the CSI driver, no Secret resource is created, alleviating the issue. Since Flant offers its customers the capability to use Certificate resources,

Reviewer 3: ```suggestion
It is possible to use a
[`csi-driver`](https://github.com/cert-manager/csi-driver) to circumvent
the problem of "too many ephemeral Secret resources stored in etcd". Using
a CSI driver, no Secret resource is created, alleviating the issue. Since Flant offers its customers the capability to use Certificate resources,
```

and wants to keep supporting the Certificate type, switching from Certificate

> resources to the CSI driver isn't an option.

Reviewer 3: ```suggestion
resources to a CSI driver isn't an option.
```

> field extensible.

When changing the value of the field `deletionPolicy` from `Delete` to `Orphan`,
the associated Secret resource immediately loses its owner reference. The user
doesn't need to wait until the certificate is renewed.
Along with this new field, we propose to deprecate the flag `--enable-certificate-owner-ref`
and introduce the new flag `--default-secret-deletion-policy`. Its values are as follows:

- When `--default-secret-deletion-policy` is set to `Orphan`, the Certificate resources
  that don't have the `deletionPolicy` field set will have their associated Secret
  resources updated (i.e., the owner reference gets removed) on the next issuance of
  the Certificate.
- When `--default-secret-deletion-policy` is set to `Delete`, the Certificate resources
  that don't have the `deletionPolicy` field set will have their associated Secret
  resources updated (i.e., the owner reference gets added) on the next issuance of
  the Certificate.
  
The effect of changing `--default-secret-deletion-policy` from `Orphan` to `Delete`
or from `Delete` to `Orphan` is not immediate: the change requires a re-issuance
of the Certificate resources.

> The default value for `--default-secret-deletion-policy` is `Orphan`.

Reviewer 3: I'd be ok with this being a behavior change to `Delete` as the default.

<!--

Can the feature be disabled after having been enabled?

Consider whether any additional steps will need to be taken to start/stop using this feature, i.e change existing resources that have had new field added for the feature before disabling it.

Do the test cases cover both the feature being enabled and it being disabled (where relevant)?

-->

### Does this feature depend on any specific services running in the cluster?

No.

### Will enabling / using this feature result in new API calls (i.e to Kubernetes apiserver or external services)?

No.

### Will enabling / using this feature result in increasing size or count of the existing API objects?

No.

### Will enabling / using this feature result in significant increase of resource usage? (CPU, RAM...)

No.

## Drawbacks

## Alternatives

It is possible to use the
[`csi-driver`](https://github.com/cert-manager/csi-driver) to circumvent
the problem of "too many ephemeral Secret resources stored in etcd". Using
the CSI driver, no Secret resource is created, alliviating the issue.

Flant relies on Certificate resources to provide X.509 certificates for

> their customers. The certificates cannot be tied to a Pod workload.

Author: @RomanenkoDenys Did I understand correctly the reason why you can't use [`cert-manager/csi-driver`](https://github.com/cert-manager/csi-driver)?

Reviewer 2: Hi! We can't use cert-manager/csi-driver because our customers have tons of different deploys with certificates. Our main goal is to do things that don't affect customers.

Reviewer 2: Usage of [cert-manager/csi-driver](https://github.com/cert-manager/csi-driver) will be good solution in future, we will discuss it in. our team.

Author: I have rephrased the paragraph to say that Certificate is part of the "user-facing API" that Flant offers to customers, meaning that entirely moving to csi-driver isn't possible. Does it make sense?

Reviewer 2: It's not a part of our API, but users knows how to use certificate resources and uses them. And they do not know anything about usage of csi-driver, so that solution complicates user's experience.

  - [Test Plan](#test-plan)
  - [Graduation Criteria](#graduation-criteria)
  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
  - [Supported Versions](#supported-versions)
- [Production Readiness](#production-readiness)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Release Signoff Checklist

This checklist contains actions which must be completed before a PR implementing this design can be merged.

- [ ] This design doc has been discussed and approved
- [ ] Test plan has been agreed upon and the tests implemented
- [ ] Feature gate status has been agreed upon (whether the new functionality will be placed behind a feature gate or not)
- [ ] Graduation criteria is in place if required (if the new functionality is placed behind a feature gate, how will it graduate between stages)
- [ ] User-facing documentation has been PR-ed against the release branch in [cert-manager/website]

## Summary

cert-manager has the ability to set the owner reference field in generated Secret resources. The option is global, and takes the form of the flag `--enable-certificate-owner-ref` set in the cert-manager controller Deployment resource.

> Let us take an example of Certificate resource:

Reviewer 3: ```suggestion
Let us take an example Certificate resource:
```

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: cert-1
  namespace: ns-1
  uid: 1e0adf8
spec:
  secretName: cert-1
```

> When `--enable-certificate-owner-ref` is passed to the cert-manager controller, when issuing the X.509 certificate, cert-manager will create a Secret resource that looks like this:

Reviewer 3: ```suggestion
When `--enable-certificate-owner-ref` is passed to the cert-manager controller, cert-manager, when issuing the X.509 certificate, will create a Secret resource that looks like this:
```

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: cert-1
  namespace: ns-1
  ownerReferences:
    - controller: true
      blockOwnerDeletion: false
      uid: 1e0adf8
      name: cert-1
      kind: Certificate
      apiVersion: cert-manager.io/v1
data:
  tls.crt: "..."
  tls.key: "..."
  ca.crt: "..."
```

> The proposition is to add a new field `certificateOwnerRef` to the Certificate resource:

Reviewer 3: ```suggestion
The proposition is to add a new field `cleanupPolicy` to the Certificate resource:
```

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
spec:
  secretName: cert-1
  cleanupPolicy: [Delete|Never] # ✨
```

The new field `cleanupPolicy` has three possible values:

> 1. When "empty", the behavior will default to not creating an owner reference on the Secret resource, unless `--enable-certificate-owner-ref` is passed.

Reviewer 4: Does this mean we don't default this field to the value of the flag during admission, and instead the controller infers the behaviour based on the flag value? Perhaps having empty automatically default to something like "Inherit" or "Default" would be clearer from a user perspective (when running a 'get -o yaml', they'll see the value of this field and be able to infer that the behaviour is based on the flag configuration).

Additionally, if a user changes the cleanupPolicy (either explicitly configured on the Certificate OR a change in the flag value) *after* issuance, will the secret reconciliation controller add (or conversely, remove) the owner reference automatically?

Reviewer 2: If user changes cleanup policy controller do not change existing secrets. Owner reference is applied on the next certificate issuance.

Author: > Perhaps having empty automatically default to something like "Inherit" or "Default" would be clearer from a user perspective

I agree, the "empty" value is often avoided in Kubernetes APIs. Let's go with `Inherit` then.

> Will the secret reconciliation controller add (or conversely, remove) the owner reference automatically?

After looking at [`SecretOwnerReferenceManagedFieldMismatch`](https://github.com/cert-manager/cert-manager/blob/f058c815a379afc36e9abff21454fc3423e40f78/internal/controller/certificates/policies/checks.go#L474), it seems like adding the flag `--enable-certificate-owner-ref` when you already have existing Secret resource will add the owner reference.

2. When `Delete`, the default behavior as described in the "empty" case is overridden and the owner reference is always created on the Secret resource.
3. When `Never`, the default behavior as described in the "empty" case is overridden and the owner reference is never created on the Secret resource.

> At first, the proposed field was named `certificateOwnerRef` and was a
> nullable boolean. James Munnelly reminded us that the Kubernetes API
> never uses boolean fields, and instead uses the string type with
> "meaningful values". On top of being more readable, it also makes the
> field extensible.

## Use-cases

> Flant manages certificates for users, and has hit a Kubernetes apiserver limitation where too many left-over Secret resources were slowing the apiserver down. This issue has happened because Certificate resources are created using auto-generated names, and Certificate resources are often deleted shortly after being created.

Reviewer 3: Could you provide a link for Flant?

> Flant manages certificates for users, and has hit a Kubernetes apiserver limitation where too many left-over Secret resources were slowing the apiserver down. This issue has happened because Certificate resources are created using auto-generated names, and Certificate resources are often deleted shortly after being created.

Reviewer 3: ```suggestion
Flant manages certificates for users, and has hit a Kubernetes apiserver limitation where too many leftover Secret resources were slowing the apiserver down. This issue has happened because Certificate resources are created using auto-generated names, and Certificate resources are often deleted shortly after being created.
```

## Questions

<!--
This section is important for producing high-quality, user-focused
documentation such as release notes.

A good summary is probably around a paragraph in length.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md

> -->

Reviewer 3: Should this be empty?

Author: I don't know what to do with empty sections yet.

-->

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
What is the desired outcome and how do we measure success?
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation- those should go into "Design Details" below.
-->

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this proposal gets implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

#### Story 2

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go into as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes/PKI ecosystem.

-->

## Design Details

> cert-manager would have to changed in a few places.

Reviewer 3: ```suggestion
cert-manager would have to change in a few places.
```

**Mutating webhook**

We propose to have no "value defaulting" for `cleanupPolicy` because the
"empty" value has a meaning for us: when `cleanupPolicy` is empty, the
presence or not of the flag `--enable-certificate-owner-ref` takes over.
To give more context, some other resources, such as the Pod resource,
will mutate the object when the value is "empty", for example the

> `imagePullPolicy` value gets defaulted to `IfNotPresent`.

Reviewer 3: ```suggestion
`imagePullPolicy` value will default to `IfNotPresent`.
```

-->

### How can this feature be enabled / disabled for an existing cert-manager installation?

<!--

Can the feature be disabled after having been enabled?

Consider whether any additional steps will need to be taken to start/stop using this feature, i.e change existing resources that have had new field added for the feature before disabling it.

Do the test cases cover both the feature being enabled and it being disabled (where relevant)?

-->

### Does this feature depend on any specific services running in the cluster?

No.

### Will enabling / using this feature result in new API calls (i.e to Kubernetes apiserver or external services)?

No.

### Will enabling / using this feature result in increasing size or count of the existing API objects?

No.

### Will enabling / using this feature result in significant increase of resource usage? (CPU, RAM...)

No.

## Drawbacks

## Alternatives

It is possible to use the
[`csi-driver`](https://github.com/cert-manager/csi-driver) to circumvent
the problem of "too many ephemeral Secret resources stored in etcd". Using

> the CSI driver, no Secret resource is created, alliviating the issue.

Reviewer 3: ```suggestion
the CSI driver, no Secret resource is created, alleviating the issue.
```

- When `--enable-certificate-owner-ref` is not passed (or is set to false), the existing
  Secret resources that have an owner reference are not changed even after a re-issuance.
  With `--default-secret-cleanup-policy` and given that `cleanupPolicy` is not set, the
  behavior is slightly different: unlike with the old flag, the existing Secret resources
  will have their owner references removed.
- When `--enable-certificate-owner-ref` is set to true, the behavior is the same as
  when `--default-secret-cleanup-policy` is set to `OnDelete` and `cleanupPolicy` is not
  set.  

The deprecated flag `--enable-certificate-owner-ref` keeps precendence over the new flag
in order to keep backwards compatibility.

When upgrading to the new flag, users can refer to the following table:

| If... | then they should replace it with... |
|-----|-----|
| `--enable-certificate-owner-ref` not passed to the controller | No change needed |
| `--enable-certificate-owner-ref=false` | Replace with `--default-secret-cleanup-policy=Never` |
| `--enable-certificate-owner-ref=true` | Replace with `--default-secret-cleanup-policy=OnDelete` |

## Use-cases

> [Flant](https://flant.com) manages certificates for users, and has hit a Kubernetes apiserver limitation where too many leftover Secret resources were slowing the apiserver down. This issue has happened because Certificate resources are created using auto-generated names, and Certificate resources are often deleted shortly after being created.

Reviewer 2: @maelvls We have clients with thousands of certificates in dev-namespaces, it is reasonable to remove secrets along with certificates by enabling --enable-certificate-owner-ref=true. On the other hand, there are certificates with secrets in the same cluster which cannot be removed, because restarting the ingress associated with these secrets violates the SLA. For these secrets, it makes sense to disable --enable-certificate-owner-ref=false.
In the current cert-manager implementation, you cannot do this.

> [Flant](https://flant.com) manages certificates for users, and has hit a Kubernetes apiserver limitation where too many leftover Secret resources were slowing the apiserver down. This issue has happened because Certificate resources are created using auto-generated names, and Certificate resources are often deleted shortly after being created.

Author: Thanks. I read https://github.com/deckhouse/deckhouse/pull/1601 and I now have a better understanding.

I also came across https://github.com/cert-manager/cert-manager/issues/3884. The author of this issue wants their Vault-issued certificates to add the owner reference on Secrets, but they don't want their Let's Encrypt-issued certificate to add the owner reference. I asked them what was their use-case.

# Design: Per-Certificate Secret Owner Reference

> 🌟 This design document was written by Maël Valais on 20 July 2022 in order to facilitate Denis Romanenko's feature request presented in [#5158](https://github.com/cert-manager/cert-manager/pull/5158).

cert-manager has the ability to set the owner reference field in generated Secret resources. The option is global, and takes the form of the flag `--enable-certificate-owner-ref` set in the cert-manager controller Deployment resource.

Let us take an example of Certificate resource:

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: cert-1
  namespace: ns-1
  uid: 1e0adf8
spec:

>   secretRef: cert-1

Reviewer 5: ```suggestion
  secretName: cert-1
```

```

When `--enable-certificate-owner-ref` is passed to the cert-manager controller, when issuing the X.509 certificate, cert-manager will create a Secret resource that looks like this:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: cert-1
  namespace: ns-1
  ownerReferences:
    - controller: true
      blockOwnerDeletion: false
      uid: 1e0adf8
      name: cert-1
      kind: Certificate
      apiVersion: cert-manager.io/v1
data:
  tls.crt: "..."
  tls.key: "..."
  ca.crt: "..."
```

The proposition is to add a new field `certificateOwnerRef` to the Certificate resource:

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
spec:

>   secretRef: cert-1

Reviewer 5: ```suggestion
  secretName: cert-1
```

  certificateOwnerRef: true  # ✨
```

It has three possible values:

1. When "empty", the behavior will default to not creating an owner reference on the Secret resource, unless `--enable-certificate-owner-ref` is passed.
2. When `true`, the default behavior as described in the "empty" case is overridden and the owner reference is always created on the Secret resource.
3. When `false`, the default behavior as described in the "empty" case is overridden and the owner reference is never created on the Secret resource.

> > **⁉️ Question:** the field name `certificateOwnerRef` does not reflect the behavior that it aims to enable. A more appropriate, less confusing name could be found, e.g., `deleteSecretUponDeletion`.

Author: As @munnerz suggested in https://github.com/cert-manager/cert-manager/pull/5158#issuecomment-1189056913, the current boolean field `certificateOwnerRef` poses two questions:

1. Do we think it makes sense to have an "open string" because we think that there is a chance that we will need to extend this in the future?
2. Is the name good enough to convey the behavior of that field?

For context, James suggests:

> Something more like:
> ```yaml
> cleanupPolicy: [Delete|Never]
> ```
> may feel more natural and be more informative to our users, as well as potentially allowing us to extend the different policies we offer in future too.

Reviewer 4: I'd also call out the Kubernetes API conventions doc which does note that bool typed fields should be avoided generally, in favour of this sort of enum value 😊

Author: Ah, I wasn't aware of this Kubernetes API convention, and now that you mention it, I realize that I never met any boolean in the Kubernetes API. Good catch.

I am in favor of following James's advice on this. I will update the design document to reflect this, and also mention whether or not a default value should be set when no value is given. I guess we should not default anything.

Reviewer 4: The default should match today's behaviour (Never) - we can set this via openapi I think 😊

Reviewer 2: We cannot set default value because this field have 3 states, "set ownerRef field", "do not set ownerRef field" and "this field do not have value and we use global setting from cert-manager"

## Use-cases

Flant manages certificates for users, and has hit a Kubernetes apiserver limitation where too many left-over Secret resources were slowing the apiserver down. This issue has happened because Certificate resources are created using auto-generated names, and Certificate resources are often deleted shortly after being created.

> 

Reviewer 5: Some questions from me:

- Are we expecting this to be made GA straight away, and if not, what is the graduation path and who has ownership of it?
- I assume we are going to add this to the `postIssuancePolicyChain`? I don't see that in the PR currently. Can we add a note that we are going to do this and include it in testing.
- What happens when I upgrade or downgrade cert-manager?
- Has the [`csi-driver`](https://github.com/cert-manager/csi-driver) been considered? If so, why is that not a good enough alternative to the use case?

Author: I tried answering the four questions in 8de5f92d2.

kind: Certificate
spec:
  secretName: cert-1
  cleanupPolicy: [OnDelete|Never] # ✨ Can be left empty.
```

The new field `cleanupPolicy` has three possible values:

1. When not set, the value set by `--default-secret-cleanup-policy` is inherited.
2. When `OnDelete`, the owner reference is always created on the Secret resource.
3. When `Never`, the owner reference is never created on the Secret resource.

> At first, the proposed field was named `certificateOwnerRef` and was a
> nullable boolean. James Munnelly reminded us that the Kubernetes API
> never uses boolean fields, and instead uses the string type with
> "meaningful values". On top of being more readable, it also makes the
> field extensible.

When changing the value of the field `cleanupPolicy` from `OnDelete` to `Never`,
the associated Secret resource immediately loses its owner reference. The user

> doesn't need to wait until the certificate is renewed. Similarly, when `cleanupPolicy`

Reviewer 3: How is this similarly statement different from the beginning of this paragraph?
